### PR TITLE
Introduce a more robust parameter system.

### PR DESCRIPTION
**Pull request by ozmodiar**
I want to propose a more robust and clear way to add parameters to the message and recipient. Doing it this way gives the developer a more clear image of which parameters each channel is going to use. Also the channel can check for the needed parameters much more easily.

2 downsides:
- It is possible that there will be duplicate data in the different parameterbags. But I think this is only a small sacrafice for the advantages you gain with this patch.
- You have to set a channel identifier in the Channel class AND in the ParameterBag class(es). If you know a better solution for this, let me know.

After applying this patch, you can use notifier like this:

``` php
    $message = new Message(new InformationType());
    $message->addParameterBag(new MailMessageParameterBag('This is the subject', 'Body.....'));

    $recipient = new Recipient();
    $recipient->addParameterBag(new MailRecipientParameterBag('someone@example.com'));

    $notifier = new Notifier(new RecoverPasswordChannelResolver());
    $notifier->sendMessage($message, array($recipient));
```

### CODE REVIEW

**Comment by NoUseFreak** [comment]
Can you add some additional tests for the bags.

It might be useful to add an example to the readme.

File: src/ParameterBag/NotifierParameterBagInterface.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * (c) Dries De Peuter <dries@nousefreak.be>
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Notifier\ParameterBag;
+
+interface NotifierParameterBagInterface {
+
+  /**
```
Comment (NoUseFreak) [inline src/ParameterBag/NotifierParameterBagInterface.php]: Please use sf2 coding standards. 
Check out php-cs-fixer to simplify this process.

File: tests/Stubs/ParameterBag.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+class ParameterBag implements NotifierParameterBagInterface {
+    /**
+     * @return string
+     */
+    public function getIdentifier()
+    {
+        return 'example_channel';
+    }
+}
```
Comment (NoUseFreak) [inline tests/Stubs/ParameterBag.php]: newline plz

File: src/ParameterBag/NotifierParameterBagInterface.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+   * @return string
+   */
+  public function getIdentifier();
+}
```
Comment (NoUseFreak) [inline src/ParameterBag/NotifierParameterBagInterface.php]: newline plz

**Comment by Ozmodiar** [comment]
Code style remarks fixed. I also renamed NotifierParameterBagInterface to ParameterBagInterface. Made more sense to me...

**Comment by NoUseFreak** [comment]
I agree. Can you provide additional tests for the ParameterBags. 

Testing if the response is null for example.

Additionally I would suggest creating a hasParameterBag instead of checking for the response. This will improve create a less type sensitive API.

**Comment by Ozmodiar** [comment]
Done... I'll leave the commits separate until the code review is done and rebase afterwards. 

Another thought: Message and Recipient seem to share a lot of code now. Do you think it is worth it to move that code to a higher level?

**Comment by NoUseFreak** [comment]
> Another thought: Message and Recipient seem to share a lot of code now. Do you think it is worth it to move that code to a higher level? 

I agree some code can be shared. But as they are not at all related, the only way to share logic is by using [traits](http://php.net/manual/en/language.oop5.traits.php).
This would require setting the minimum PHP requirement to `~5.4` which is not a problem as `5.3` is [EOL](http://php.net/supported-versions.php).

**Comment by Ozmodiar** [comment]
Indeed, this looks like a good use case for traits. I've implemented a ParameterBagTrait... You can check it out.

**Comment by NoUseFreak** [comment]
Looks good. After a rebase it will be merged.
Thank you very much for this contribution!

**Comment by Ozmodiar** [comment]
Ok, done...

**Pull request merged**